circuit/types/integers/src/mul_checked.rs

```diff
@@ -119,97 +113,101 @@ impl<E: Environment, I: IntegerType> MulChecked<Self> for Integer<E, I> {
             Self::ternary(operands_same_sign, &product, &Self::zero().sub_wrapped(&product))
         } else {
-            // Compute the product of `self` and `other`.
-            let (product, carry) = Self::mul_with_carry(self, other);
-
-            // For unsigned multiplication, check that none of the carry bits are set.
-            for bit in carry.iter() {
-                E::assert_eq(bit, E::zero());
-            }
-
-            // Return the product of `self` and `other`.
-            product
+            // Compute the product of `self` and `other`, checking for overflow.
+            Self::mul_and_check(self, other)
         }
     }
 }
 
 impl<E: Environment, I: IntegerType> Integer<E, I> {
-    /// Multiply the integer bits of `this` and `that` in the base field,
-    /// returning the carry bits in a separate vector.
-    /// The returned carry bits are not quite the ones whose value is the carry:
-    /// instead, they are the concatenation of the carry bits from two sub-multiplications
-    /// (see the comments in the code for details);
-    /// this is adequate for the current uses of this function,
-    /// where the callers just use the returned carry bits to check that they are all zero.
+    /// Multiply the integer bits of `this` and `that`, checking for overflow.
+    /// This function assumes that `this` and `that` are non-negative.
     #[inline]
-    pub(super) fn mul_with_carry(this: &Integer<E, I>, that: &Integer<E, I>) -> (Integer<E, I>, Vec<Boolean<E>>) {
+    fn mul_and_check(this: &Integer<E, I>, that: &Integer<E, I>) -> Integer<E, I> {
         // Case 1 - 2 integers fit in 1 field element (u8, u16, u32, u64, i8, i16, i32, i64).
         if 2 * I::BITS < (E::BaseField::size_in_bits() - 1) as u64 {
-            // Instead of multiplying the bits of `self` and `other` directly, the integers are
-            // converted into field elements, multiplied, and the result is converted back to an integer.
-            // Note: This is safe as the field is larger than the maximum integer type supported.
-            let product = (this.to_field() * that.to_field()).to_lower_bits_le(2 * I::BITS as usize);
+            // Instead of multiplying the bits of `self` and `other`, witness the integer product.
+            let product: Integer<E, I> = witness!(|this, that| this.mul_wrapped(&that));
 
-            // Split the integer bits into product bits and carry bits.
-            let (bits_le, carry) = product.split_at(I::BITS as usize);
+            // Check that the computed product is equal to witnessed product, in the base field.
+            // Note: The multiplication is safe as the field twice as large as the maximum integer type supported.
+            let computed_product = this.to_field() * that.to_field();
+            let witnessed_product = product.to_field();
+            E::assert_eq(&computed_product, &witnessed_product);
```

Comment:
I think the 3 lines above generate 2 constraints (one for computed_product, one for assert_eq) and 1 variable (computed_product). The following should generate a single constraint and no variable:
```
E::enforce(|| (this.to_field(), that.to_field(), product.to_field()))
```